src/components/useQueryMutator.tsx

```diff
@@ -0,0 +1,73 @@
+import { useEffect, useState } from "react";
+import { useDebounce } from "use-debounce/lib";
+import { useQueryParams } from "use-query-params";
+import {DecodedValueMap, QueryParamConfigMap } from 'serialize-query-params';
+import { buildQueryParamsMutators } from "../lib/queryParams";
+import { ICreatureIntFilterSchema, ICreatureModel, ICreaturesSearchSchema, ICreatureStrFilterSchema } from "../lib/openAPI";
+import FilterDrawer from "./filters/FilterDrawer";
+import FilterButtons from "./filters/FilterButtons";
+import SearchInput from "./SearchInput";
+import { TablePagination } from "@mui/material";
+
+export default function useQueryParamMutator(QPCMap: QueryParamConfigMap, filterFields, fetcher: (arg0: DecodedValueMap<QueryParamConfigMap>) => Promise<any>) {
+    const [results, setResults] = useState<ICreatureModel[]>([]);
+    const [count, setCount] = useState<number>(0);
+  
+    const [isFilterDrawerOpen, setIsFilterDrawerOpen] = useState<boolean>(false);
+    const [query, setQuery] = useQueryParams(QPCMap);
+    const [queryDebounced] = useDebounce(query, 200);
+
+    useEffect(() => {
+        fetcher(queryDebounced).then((response: ICreaturesSearchSchema) => {
+          if (response.pagination) {
+            const {
+              data,
+              pagination: { count },
+            } = response;
+            setResults(data);
+            setCount(count);
+          } else {
+            // TODO: handle validation error
+          }
+        });
+      }, [queryDebounced]);
+
+      const filterProps = buildQueryParamsMutators<
+        ICreatureStrFilterSchema | ICreatureIntFilterSchema
+      >(query, setQuery);
+
+      const filterButtonsProps = {
+        hasFilters: query.filters.length ? true : false,
+        setIsFilterDrawerOpen: setIsFilterDrawerOpen,
+        clearFilters: filterProps.clearFilters,
+      };
+
+      const searchInputProps = {
+        q: query.q,
+        qChange: filterProps.qChange,
+      }
+
+      const filterDrawerProps = {
+          ...filterProps,
+          isFilterDrawerOpen,
+          filters: query.filters,
+          fields: filterFields,
+          setIsFilterDrawerOpen,
+      }
+
+      const tablePaginationProps = {
+            role: "presentation",
+            count: count,
+            page: query.page,
+            labelRowsPerPage: "Num: ",
+            onPageChange: filterProps.pageChange,
+            rowsPerPage: query.size,
+            onRowsPerPageChange: filterProps.sizeChange,
+      };
+
+
+      return {results, query, filterButtonProps: filterButtonsProps, FilterButtons, filterProps, filterDrawerProps,FilterDrawer,
```

Comment:
Lets just return everything as top level. We don't want to couple this hook too tightly to the UI implementation. 

There is also some worry about passing these special props objects along instead of specific props as hides needed context in a hook.

Reply:
I may be misunderstanding what top level means?

There is still an options do
```javascript
const top = useQueryParamMutator(queryParamsStructure, FIELDS, fetchCreatures);
// use filterButtonProps somehow
top.filterButtonProps;
```
Another way would be to create our own Table component and add optional functionalities.
https://codesandbox.io/s/compount-components-with-a-hook-txolo?file=/table/table.hooks.js